test/State.hs

```diff
@@ -25,45 +25,45 @@ tests = testGroup "State"
     [ testMonad
     , testMonadFix
     , testState
-    ] >>= ($ RunC LazyStateC.runState)
+    ] >>= ($ runC LazyStateC.runState)
   , testGroup "StateC (Strict)" $
     [ testMonad
     , testMonadFix
     , testState
-    ] >>= ($ RunC StrictStateC.runState)
-  , testGroup "StateT (Lazy)"   $ testState (RunC (fmap (fmap swap) . flip LazyStateT.runStateT))
-  , testGroup "StateT (Strict)" $ testState (RunC (fmap (fmap swap) . flip StrictStateT.runStateT))
-  , testGroup "RWST (Lazy)"     $ testState (RunC (runRWST LazyRWST.runRWST))
-  , testGroup "RWST (Strict)"   $ testState (RunC (runRWST StrictRWST.runRWST))
+    ] >>= ($ runC StrictStateC.runState)
+  , testGroup "StateT (Lazy)"   $ testState (runC (fmap (fmap swap) . flip LazyStateT.runStateT))
+  , testGroup "StateT (Strict)" $ testState (runC (fmap (fmap swap) . flip StrictStateT.runStateT))
+  , testGroup "RWST (Lazy)"     $ testState (runC (runRWST LazyRWST.runRWST))
+  , testGroup "RWST (Strict)"   $ testState (runC (runRWST StrictRWST.runRWST))
   ] where
-  testMonad    run = Monad.test    (m (gen s)) a b c (atom "(,)" (,) <*> s <*> unit) run
-  testMonadFix run = MonadFix.test (m (gen s)) a b   (atom "(,)" (,) <*> s <*> unit) run
-  testState    run = State.test s  (m (gen s)) a                                     run
+  testMonad    run = Monad.test    (m (gen0 s) (\ _ _ -> [])) a b c (pair <*> s <*> unit) run
+  testMonadFix run = MonadFix.test (m (gen0 s) (\ _ _ -> [])) a b   (pair <*> s <*> unit) run
+  testState    run = State.test    (m (gen0 s) (\ _ _ -> [])) a               s           run
   runRWST f s m = (\ (a, s, ()) -> (s, a)) <$> f m s s
 
 
-gen
-  :: forall s m sig
+gen0
+  :: forall s m a sig
   .  (Has (State s) sig m, Arg s, Show s, Vary s)
   => Gen s
-  -> GenM m
-  -> GenM m
-gen s _ a = choice
+  -> Gen a
+  -> [Gen (m a)]
+gen0 s a =
   [ label "gets" (gets @s) <*> fn a
   , infixL 4 "<$" (<$) <*> a <*> (label "put" put <*> s)
   ]
 
 
 test
   :: (Has (State s) sig m, Arg s, Eq a, Eq s, Show a, Show s, Vary s)
-  => Gen s
-  -> GenM m
+  => GenM m
   -> Gen a
-  -> RunC s ((,) s) m
+  -> Gen s
+  -> Run ((,) s) ((,) s) m
```

Comment:
I haven’t yet generalized this one over the input state, because we don’t need it urgently. It can wait for some (interesting!) state/nondet tests, or something.